### PR TITLE
Update documuent viewer in the attachment widget on attribute change

### LINKED ISSUE #37678: attachment widget: expression to control integrated document viewer should react immediately after user selected a file
**Feature description.**
In the attachment widget, there is an expression to control "NoContent|Image|Web" view. This expression is currently only evaluated on the form loading.

image

It would be nice if a signal could be connected to the widget where the user selects a file so that the expression could fire after the user selected a new file.

image

Example expression:

```
CASE WHEN filepath ILIKE '%.jpg' THEN 'Image'
	 ELSE 'NoContent'
END
```

### PR DESCRIPTION
**Pull request by signedav**
In the Attachment-Widget the document viewer type can be controlled over conditions. 

On update of an attribute in the form, the result of this condition is reevaluated and the document viewer type is updated on runtime.

If you set this expression:
```
CASE WHEN  "infofile"  ILIKE '%.jpg' AND "show_preview" 
THEN 
  'Image'
ELSE 
  CASE WHEN  "infofile"  ILIKE 'http%' AND "show_preview" 
  THEN 
    'Web'
  ELSE 
    'NoContent'
  END
END
```
Then you will get this behavior:
document_viewer

On change of an attribute the documentViewer expression of the externalRessourceWidget depends on, the documentViewer should be updated.

To not be needed to check for all the widget values in the form, the mCurrentFormFeature is introduced that contains all the values that are received on change of the widget values.

Resolves #37678

### CODE REVIEW

File: python/gui/auto_generated/qgsattributeform.sip.in [hunk 1]
```diff
@@ -44,6 +44,14 @@ class QgsAttributeForm : QWidget
 
     const QgsFeature &feature();
 
+    const QgsFeature &currentFormFeature();
```
Comment (m-kuhn) [inline python/gui/auto_generated/qgsattributeform.sip.in:47]:
```suggestion
    QgsFeature currentFormFeature() const;
```

Better safe than sorry, returning refs has potential risks if used unconsciously and returning an implicitly shared construct like a feature is almost free.

Comment (signedav) [reply python/gui/auto_generated/qgsattributeform.sip.in:47]: Okay.

**Comment by m-kuhn** [comment]
Thanks @signedav 

@elpaso what do you think about keeping a "currentFormFeature" up to date this way?
I think we already have code to do the same in various places (e.g. for child forms, for drill down cascading, maybe more). Having it once in a centralized way sounds like the right thing to do?

File: src/gui/qgsexternalresourcewidget.h [hunk 2]
```diff
@@ -143,6 +143,11 @@ class GUI_EXPORT QgsExternalResourceWidget : public QWidget
      */
     void setDefaultRoot( const QString &defaultRoot );
 
+    /**
+     * Reloads the document from the path of the filewidget
```
Comment (m-kuhn) [inline src/gui/qgsexternalresourcewidget.h:147]: since 3.16 ?

File: src/gui/qgsattributeform.h [hunk 3]
```diff
@@ -75,6 +75,14 @@ class GUI_EXPORT QgsAttributeForm : public QWidget
 
     const QgsFeature &feature() { return mFeature; }
 
+    /**
+     * Returns the feature that is currently displayed in the form with all
+     * the changes received on editing the values in the widgets.
+     *
+     * \since QGIS 3.16
+     */
+    const QgsFeature &currentFormFeature() { return mCurrentFormFeature; }
```
Comment (m-kuhn) [inline src/gui/qgsattributeform.h:84]: also needs updating here and in the header

**Review by m-kuhn — APPROVED** [review]

File: src/gui/editorwidgets/qgsexternalresourcewidgetwrapper.cpp [hunk 4]
```diff
@@ -112,15 +112,25 @@ void QgsExternalResourceWidgetWrapper::setFeature( const QgsFeature &feature )
           dvc = QgsExternalResourceWidget::NoContent;
         }
         mQgsWidget->setDocumentViewerContent( dvc );
+        mQgsWidget->reloadDocument();
```
[the remark this answers is not in the document]
Comment (signedav) [reply src/gui/editorwidgets/qgsexternalresourcewidgetwrapper.cpp:115]: Hm... Yes, we could do that. Means I would remove `reloadDocument()` completely and call `loadDocument(path)` in the `setDocumentViewerContent`.

[the remark this answers is not in the document]
Comment (m-kuhn) [reply src/gui/editorwidgets/qgsexternalresourcewidgetwrapper.cpp:115]: Less risk of forgetting it after calling `setDocumentViewerContent`?

[the remark this answers is not in the document]
Comment (signedav) [reply src/gui/editorwidgets/qgsexternalresourcewidgetwrapper.cpp:115]: Yes.

**Pull request merged**